File: LeetCodePractice/src/twoPointerSlidingWindow/_0007_NumberOfSubArraysOfSizeKAndAverageGreaterThanOrEqualToThreshold.java
package twoPointerSlidingWindow;

public class _0007_NumberOfSubArraysOfSizeKAndAverageGreaterThanOrEqualToThreshold {

	public int numOfSubarrays(int[] arr, int k, int threshold) {
		int currSum = 0; // For storing the sum value of window size
		int countK = 0; // To check if the size is achieved then start adding the values
		int countRes = 0; // To store the result count
		int len = arr.length;
		int avg;
		for (int i = 0; i < len; i++) {

			currSum += arr[i];
			countK++;
			if (countK >= k) {
				avg = (currSum / k);
				if (avg >= threshold) {
					countRes++;
				}
				/**
				 * This will subtract the first element in the window and in the next iteration,
				 * the new window will be added. For e.g: arr=[1,2,3,4,5] and k = 3 i=2 currSum
				 * = (1+2+3) = 6 currSum = currSum - arr[(i+1)-k]; window is (1,2,3) - Subtract
				 * 1 from the window then shift the window. currSum = 6 - arr[0];
				 **/
				currSum -= arr[(i + 1) - k];
			}
		}
		return countRes;
	}

}

File: LeetCodePython/zzzz_TwoPointer&SlidingWindow/_0007_NumberofSub-arraysOfSizeKAndAverageGreaterThanOrEqualToThreshold.py
from typing import List
class Solution:
    # approach - 1
    def numOfSubarrays(arr, k, threshold):
        target_sum = k * threshold
        current_sum = sum(arr[:k])

        if current_sum >= target_sum:
            count = 1
        else:
            count = 0 
    
        for i in range(k, len(arr)):
            current_sum += arr[i] - arr[i - k]
            if current_sum >= target_sum:
                count += 1
        return count

# approach - 2
    def numOfSubarrays(arr, k, threshold):
        curr_sum = 0  # To store the sum of the current window
        count_k = 0  # To track the window size
        count_res = 0  # To store the result count
        length = len(arr)
        
        for i in range(length):
            curr_sum += arr[i]
            count_k += 1
            
            # Check if the window size is at least k
            if count_k >= k:
                avg = curr_sum / k
                if avg >= threshold:
                    count_res += 1
                    
                # Shift the window by subtracting the first element of the current window
                curr_sum -= arr[i + 1 - k]
        
        return count_res

    def numOfSubarrays(self, arr: List[int], k: int, threshold: int) -> int:
        curr_sum = 0  # To store the sum of the current window
        count_k = 0  # To track the window size
        count_res = 0  # To store the result count
        length = len(arr)
        
        for i in range(length):
            curr_sum += arr[i]
            count_k += 1
            
            # Check if the window size is at least k
            if count_k >= k:
                avg = curr_sum / k
                if avg >= threshold:
                    count_res += 1
                    
                # Shift the window by subtracting the first element of the current window
                curr_sum -= arr[i + 1 - k]
        return count_res
    
#   sliding window - 1 from neetcode.io
    def numOfSubarrays(self, arr: List[int], k: int, threshold: int) -> int:
        res = 0
        curSum = sum(arr[:k - 1])

        for L in range(len(arr) - k + 1):
            curSum += arr[L + k - 1]
            if (curSum / k) >= threshold:
                res += 1
            curSum -= arr[L]
        return res